velox/expression/CastExpr.cpp

```diff
@@ -452,6 +452,62 @@ VectorPtr CastExpr::applyRow(
       std::move(newChildren));
 }
 
+template <typename FROM, typename T>
+void applyDecimalCastKernel(
+    const SelectivityVector& rows,
+    DecodedVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType,
+    VectorPtr castResult) {
+  auto sourceVector = input.base()->as<SimpleVector<FROM>>();
+  T* castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<T>>()->mutableRawValues();
+  const auto& fromPrecisionScale = getDecimalPrecisionScale(*fromType);
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+    try {
+      DecimalUtil::rescaleWithRoundUp<T>(
+          castResultRawBuffer[row],
+          sourceVector->valueAt(row).unscaledValue(),
+          fromPrecisionScale.second,
+          toPrecisionScale);
+    } catch (VeloxRuntimeError& ex) {
+      castResult->setNull(row, true);
+    }
+  });
+}
+
+template <typename T>
+VectorPtr CastExpr::applyDecimal(
+    const SelectivityVector& rows,
+    DecodedVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  VectorPtr castResult;
+  context.ensureWritable(rows, toType, castResult);
+  (*castResult).clearNulls(rows);
+  switch (fromType->kind()) {
```

Comment:
I feel we can avoid the multiple layers of templating.
Should we handle all 4 cases as in decimal arithmetic here and avoid templating `applyDecimal()`?

Reply:
We have only one template function now.